### ISSUE TITLE
Use exsiting Owin Providers

### ISSUE THREAD

**Issue by TonyAbell** [freya-fs/freya#103]
Its is possible to wire up existing owin providers into the Freya pipeline?

It would be nice to add support for the all the security modules provided by MSFT [here](http://www.asp.net/aspnet/overview/owin-and-katana/owin-oauth-20-authorization-server)

**Comment by kolektiv** [comment]
It should be possible, and where libraries expose themselves as simple "AppFunc"s then it's very easy to do. However, libraries which take the Microsoft-esque IAppBuilder approach (and only that) are currently a bit more awkward. However, it's important that they work well and simply, so I'm going to have a look at this... well, right now actually. Will update soon!

**Comment by kolektiv** [comment]
Have started an approach to this, see/watch https://github.com/kolektiv/freya/tree/feature/compatibility for progress. In essence, most of the Microsoft stuff is coupled to the `IAppBuilder` interface abstraction, and does not expose itself as OWIN spec compatible functions. This is irksome, but not surprising. One approach to this is to allow the use of the IAppBuilder approach in cases where people need compatibility with projects (middleware generally) that require it.

The new `Freya.Compatibility` provides an extension for `IAppBuilder` called `UseFreyaPipeline` which does much what you would expect - allowing you to use a Freya component in an IAppBuilder pipeline.

I've started looking at the tutorial referenced as well. Some of the functions that it requires you to implement require data not currently neatly exposed by Freya (cookies and forms, which are on the roadmap/in progress), and some require some very Microsoft abstraction specific access. It may make sense to be able to add that to `Freya.Compatibility` however.

Anyway, some progress, although I'd be happier if Microsoft hadn't made such workarounds a requirement! :)

**Comment by panesofglass** [comment]
@kolektiv, thanks for working on this. IIRC, the `Microsoft.Owin` pieces *should* all expose an build on `OwinMiddleware` and expose the standard `AppFunc` / `MidFunc` signatures. If so, then the earlier OWIN mapping should work. I know some wrappers, such as those for Web API, did not conform and would require something like this. However, I thought I remembered seeing that even the Thinktecture bits exposed the `MidFunc` or `AppFunc` signatures. Which are you seeing that don't conform?

**Comment by kolektiv** [comment]
Ah let me take a look and see whether I can. It didn't seem obvious to me last night, but I might have missed something :)

**Comment by kolektiv** [comment]
I'm not sure now, having looked. It seems like they almost might, but have hidden away that kind of thing as they all want their own configuration, etc. Would you be able to put together an example of using something like the Microsoft.Owin.Security.Cookies middleware? It seems like it should work somehow, but I'm not totally sure of the right way!

**Comment by theimowski** [comment]
FWIW I have this very middleware set up [here](https://github.com/theimowski/FreyaMusicStore/blob/master/App.fs#L36-50) - but I used the `IAppBuilder.Use` method for that

**Comment by panesofglass** [comment]
@theimowski your code highlights another thing that will be tricky for us to incorporate when running on IIS via ASP.NET: stage markers to correctly run pipeline components within the correct ASP.NET event cycle. We may have to skip that piece.

**Comment by theimowski** [comment]
TBH i haven't even checked if the stage market is mandatory there - just copied over what I saw in some implementation

**Comment by panesofglass** [comment]
It allows the ASP.NET lifecycle to correctly set the `Thread.CurrentPrincipal` in the right event stage. If you don't, it's possible your threads won't get set correctly (and therefore `HttpContext.User` may not be set correctly).

**Comment by panesofglass** [comment]
Looks like we would need to rewrite some of the `Microsoft.Owin.*.*Middleware` wrappers in order to make it work nicely. I wonder if it would be even better to just write our own. Most of the complexity is in the OO representation rather than the actual implementation.

**Comment by kolektiv** [comment]
Yeah it's interesting. I've looked at some of the content, and much of it does, as you say, seem to basically concern fiddling with state in a very verbose way. Once our typed access to cookies, for example, is complete, it feels like a lot of things could be written more readably and concisely using native Freya idioms (and, with any luck, exposed as genuinely OWIN compliant modules that anyone could take advantage of if they actually support OWIN).

**Comment by panesofglass** [comment]
@kolektiv indeed. From what I've seen, one of the hardest parts is making sure to send the correct challenge headers and model the challenge header dance correctly. I think that could be done better through F#, so I'm happy to dig into this further. I think it's worth it.

**Comment by panesofglass** [comment]
Documenting from our conversation Friday, June 26: the main missing piece here is that we need a machine extension once we finally get the middleware working. However, we are also debating the advantage of leveraging the existing Katana bits versus writing our own. Would be nice to get both options working.

**Comment by panesofglass** [comment]
If we do roll our own, we may want to try to capture what we can to share with Suave: http://blog.tamizhvendan.in/blog/2015/07/15/securing-apis-in-suave-using-json-web-token/
